### PR TITLE
11851 Enable messaging features for Jakarta EE9

### PR DESCRIPTION
**Pull request by dazavala**
Ref: #11851 

These changes enable the following Jakarta Messaging features within the io.openliberty feature namespace:

- `messaging-3.0`  (A.K.A. `jms-x.0` in JavaEE)
- `messagingServer-3.0`  (A.K.A.`wasJmsServer-1.0`)
- `messagingClient-3.0`  (A.K.A. `wasJmsClient-2.0`)
- `messagingSecurity-3.0`  (A.K.A. `wasJmsSecurity-1.0`)

Also, the changes move the Jakarta Messaging 3.0 API bundle to `io.openliberty.jakarta.messaging-3.0` to comply with the [current artifact naming conventions](https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-naming-guidelines).

The feature `messagingMdb-3.0` (A.K.A wasJmsMdb-x.y) was excluded as it would be superseded by feature `mdb-4.0` upon release.

FAT com.ibm.ws.messaging.open_fat is fully enabled to repeat for ee9.  The FAT suite verifies simple use-cases over the messagingServer and messagingClient features, which pulls messaging-3.0 into the server composition.

### CODE REVIEW

**Comment by dazavala** [comment]
#build

File: dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.messaging-3.0.internal.feature
Comment (tevans78) [inline dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.messaging-3.0.internal.feature:2]: This is an internal feature common across each of the public messaging features, right?
Does there need to be a separate private API feature (`io.openliberty.jakarta.messaging.3.0`) for other components to pull in just the API ... or should they use the api bundle directly?

Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.messaging-3.0.internal.feature:2]: Oddly no, not at this time.  Features that require the messaging api declare a dependency on this internal feature.  I can certainly add the API feature and modify the internal feature dependency.  Although the new API feature would be nearly vestigial, maintaining this feature pattern will facilitate serviceability and devOps.

Comment (tevans78) [reply dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.messaging-3.0.internal.feature:2]: That would be fine if this feature only pulled in APIs needed for Messaging... but it also pulls in connectors internals and transactions.

Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.messaging-3.0.internal.feature:2]: No problem!  I meant "maintaining the _api-specific_ feature pattern ...".  The change is uploaded.  Note that a feature specific to the javaee/javax jms api does not exist in previous releases.

**Comment by andrea-pichler** [comment]
No message review requested from L2

File: dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties
Comment (dmuelle) [inline dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties:17]: **change**
This feature enables the configuration of resource adapters to access messaging systems using \
 the Jakarta Message Service API.
**to**
This feature enables the configuration of resource adapters to access messaging systems by using \
 the Jakarta Message Service API.

File: dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties
Comment (dmuelle) [inline dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties:18]: `This also includes the configuration of Jakarta Message Service connection \
 factories, queues, topics and activation specifications.`

What does "This" refer to ? The feature?

File: dev/com.ibm.websphere.appserver.features/visibility/public/messagingClient-3.0/resources/l10n/io.openliberty.messagingClient-2.0.properties
Comment (dmuelle) [inline dev/com.ibm.websphere.appserver.features/visibility/public/messagingClient-3.0/resources/l10n/io.openliberty.messagingClient-2.0.properties:20]: IBM style specifies "Integration Bus" should be capitalized

File: dev/com.ibm.websphere.appserver.features/visibility/public/messagingServer-2.0/resources/l10n/io.openliberty.messagingServer-2.0.properties
Comment (dmuelle) [inline dev/com.ibm.websphere.appserver.features/visibility/public/messagingServer-2.0/resources/l10n/io.openliberty.messagingServer-2.0.properties:18]: **change**
Applications can send and receive messages using the messagingClient feature.
**to**
Applications can send and receive messages by using the messagingClient feature.

File: dev/fattest.simplicity/src/componenttest/rules/repeater/EE8FeatureReplacementAction.java
Comment (tkburroughs) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/EE8FeatureReplacementAction.java:55]: Same changes are needed for EE7FeatureReplacementAction

Comment (tkburroughs) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/EE8FeatureReplacementAction.java:55]: Should be "jms-2.0", not 3.0

File: dev/fattest.simplicity/src/componenttest/rules/repeater/EE7FeatureReplacementAction.java
Comment (tkburroughs) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/EE7FeatureReplacementAction.java:53]: Should be "jms-2.0", not 3.0

File: dev/com.ibm.websphere.appserver.features/visibility/public/messagingClient-3.0/resources/l10n/io.openliberty.messagingClient-2.0.properties
Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/public/messagingClient-3.0/resources/l10n/io.openliberty.messagingClient-2.0.properties:20]: I'm finding that SIB terms are either all capitalized or all lower case.  All are now capitalized.

File: dev/com.ibm.websphere.appserver.features/visibility/public/messagingServer-2.0/resources/l10n/io.openliberty.messagingServer-2.0.properties
Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/public/messagingServer-2.0/resources/l10n/io.openliberty.messagingServer-2.0.properties:18]: Change made.

File: dev/fattest.simplicity/src/componenttest/rules/repeater/EE8FeatureReplacementAction.java
Comment (dazavala) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/EE8FeatureReplacementAction.java:55]: Changes made.

File: dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties
Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties]: "This" is the feature, messaging-3.0.  The verb is now changed to enables rather than includes.

File: dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties
Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties]: Change made.

**Comment by dazavala** [comment]
#build

**Review by aguibert — APPROVED** [review]

File: dev/fattest.simplicity/src/componenttest/rules/repeater/EE7FeatureReplacementAction.java
Comment (dazavala) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/EE7FeatureReplacementAction.java]: Change made.

File: dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.messaging-3.0.internal.feature
Comment (KyleAure) [inline dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.messaging-3.0.internal.feature:8]: `io.openliberty.jakarta.connectors-2.0` should be the only feature you need here as it pulls in `io.openliberty.connectors-2.0.internal`

**Comment by dazavala** [comment]
#build 
Removed the inconsistent parallel activation header from feature descriptors.

**Review by KyleAure — APPROVED** [review]

File: dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.messaging-3.0.internal.feature
Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.messaging-3.0.internal.feature:8]: Looking at the feature descriptors, `io.openliberty.jakarta.connectors-2.0` provisions the API bundle, only.

**Comment by chackoj** [comment]
The new names for the features are much better! Would it be ok to use the same version number for all of them? i.e.:—
* `messaging-3.0`
* `messagingServer-3.0`
* `messagingClient-3.0`
* `messagingSecurity-3.0`
This would be simpler and clearer. There would be no expectation for `messagingServer-2.0` to work with `wasJmsClient-2.0`, for example.

**Review by tkburroughs — APPROVED** [review]

**Comment by tkburroughs** [comment]
#run-libby-bot

File: dev/com.ibm.websphere.appserver.features/visibility/public/messagingServer-2.0/io.openliberty.messagingServer-2.0.feature
Comment (chackoj) [inline dev/com.ibm.websphere.appserver.features/visibility/public/messagingServer-2.0/io.openliberty.messagingServer-2.0.feature:2]: As mentioned elsewhere, please can we make this 3.0 instead of 2.0?

File: dev/wlp-jakartaee-transform/rules/jakarta-messaging.properties
Comment (chackoj) [inline dev/wlp-jakartaee-transform/rules/jakarta-messaging.properties:3]: I note that elsewhere in this set of changes, a file has been changed solely to add a trailing newline. By that token, does this also need one?

**Review by chackoj — DISMISSED** [review]
Only requested changes are about changing version 2.0 to version 3.0 for consistency.

File: dev/com.ibm.websphere.appserver.features/visibility/public/messagingServer-2.0/io.openliberty.messagingServer-2.0.feature
Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/public/messagingServer-2.0/io.openliberty.messagingServer-2.0.feature:2]: Yes, we've the green light on bumping all jakarta messaging features to version 3.0.

File: dev/wlp-jakartaee-transform/rules/jakarta-messaging.properties
Comment (dazavala) [reply dev/wlp-jakartaee-transform/rules/jakarta-messaging.properties:3]: We can ignore the newline.  Other file(s) indicate a newline difference because I either added or removed a newline while resolving merge conflicts.

File: dev/com.ibm.websphere.appserver.features/visibility/public/messagingSecurity-2.0/io.openliberty.messagingSecurity-2.0.feature
Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/public/messagingSecurity-2.0/io.openliberty.messagingSecurity-2.0.feature:2]: Yes we can.

File: dev/com.ibm.websphere.appserver.features/visibility/public/messagingClient-3.0/io.openliberty.messagingClient-3.0.feature
Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/public/messagingClient-3.0/io.openliberty.messagingClient-3.0.feature:4]: The dependency on private feature com.ibm.websphere.appserver.eeCompatible-9.0 will ensure that messagingClient cannot be configured with any javaEE[6..8]-compatible feature.  Same for the other messaging features.  At server startup when the feature resolver detects such a configuration it will create a singleton conflict.  The feature manager will then log the conflict as an error specific to javaee+jakartaee feature incompatibility and remove the conflicting features from list of features to provision.

**Comment by dazavala** [comment]
#build
- Bump/align the version of all jakarta messaging features to 3.0.
- Modify the jms metatype transform to replace "javax.jms" within the attribute values of config elements, only.  The previous solution would also replace "javax.jms" within the config element type names (i.e. ID values,) which introduces new config types into the ee9 server.

File: dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties
Comment (dmuelle) [inline dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties:19]: add serial comma per IBM style:

This feature also enables the configuration of Jakarta Message Service connection \
 factories, queues, topics, and activation specifications.

**Review by dmuelle — CHANGES_REQUESTED** [review]
Changes look good- one more minor edit to add a serial comma. Sorry for the delay, I was out last week but will approve promptly once the file is updated. Thanks

**Comment by dazavala** [comment]
#build

File: dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties
Comment (dazavala) [reply dev/com.ibm.websphere.appserver.features/visibility/public/messaging-3.0/resources/l10n/io.openliberty.messaging-3.0.properties]: Change made.

**Comment by dazavala** [comment]
#build
Last minute change to messaging-3.0 localizable resource.

**Comment by dazavala** [comment]
#build

**Comment by dazavala** [comment]
#run-libby-bot
Builds are completing w/o issue.  Merge pending @dmuelle's approval.

**Review by dmuelle — APPROVED** [review]
LGTM

**Pull request merged**
